### ISSUE TITLE
Runner errors for Python 3.7 when done

### ISSUE THREAD

**Issue by basnijholt** [python-adaptive/adaptive#156]
```
Exception in thread QueueManagerThread:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/opt/conda/lib/python3.7/threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/conda/lib/python3.7/concurrent/futures/process.py", line 368, in _queue_management_worker
    thread_wakeup.clear()
  File "/opt/conda/lib/python3.7/concurrent/futures/process.py", line 92, in clear
    while self._reader.poll():
  File "/opt/conda/lib/python3.7/multiprocessing/connection.py", line 255, in poll
    self._check_closed()
  File "/opt/conda/lib/python3.7/multiprocessing/connection.py", line 136, in _check_closed
    raise OSError("handle is closed")
OSError: handle is closed
```

**Comment by basnijholt** [comment]
This is actually a bug in Python 3.7: https://bugs.python.org/issue34073

**Comment by basnijholt** [comment]
Related: https://bugs.python.org/issue34075

Didn't fully read it yet, but it might have implications for `adaptive`.

**Comment by basnijholt** [comment]
The following code is a stripped down `Runner`:
```python
import random
import concurrent.futures as concurrent
import asyncio
from adaptive.runner import _ensure_executor

class RandomSampler:
    def __init__(self, function, *,
                 executor=None, ntasks=1,
                 shutdown_executor=False,
                 ioloop=None,
                 nvalues=100):

        self.executor = _ensure_executor(executor)
        self._max_tasks = ntasks
        self.shutdown_executor = shutdown_executor or (executor is None)
        self.ioloop = ioloop or asyncio.get_event_loop()
        self.task = self.ioloop.create_task(self._run())
        self.function = function
        self.result = {}
        self.nvalues = nvalues
        self.pending_points = {}

    async def _run(self):
        first_completed = asyncio.FIRST_COMPLETED

        try:
            while not len(self.result) > self.nvalues:
                futures = self._get_futures()
                done, _ = await asyncio.wait(futures,
                                             return_when=first_completed,
                                             loop=self.ioloop)
                self._process_futures(done)
        finally:
            remaining = self._remove_unfinished()
            if remaining:
                await asyncio.wait(remaining)
            self._cleanup()

    def _get_max_tasks(self):
        return self._max_tasks

    def _process_futures(self, done_futs):
        for fut in done_futs:
            x = self.pending_points.pop(fut)
            y = fut.result()
            self.result[x] = y

    def _get_futures(self):
        n_new_tasks = max(0, self._max_tasks - len(self.pending_points))

        points = [random.random() for _ in range(n_new_tasks)]

        for x in points:
            fut = self.ioloop.run_in_executor(self.executor, self.function, x)
            self.pending_points[fut] = x

        futures = list(self.pending_points.keys())
        return futures

    def _remove_unfinished(self):
        remaining = list(self.pending_points.keys())
        for fut in remaining:
            fut.cancel()
        return remaining

    def _cleanup(self):
        if self.shutdown_executor:
            self.executor.shutdown(wait=False)

def f(x):
    return x
```

```python
# error occurs
runner = RandomSampler(f, ntasks=10) 
```

```python
# Error doesn't occer
ex = concurrent.ProcessPoolExecutor()
```

**Comment by basnijholt** [comment]
I've opened an issue in the Python bug tracker: https://bugs.python.org/issue36281

**Comment by basnijholt** [comment]
Closed by #164.

This should be changed back in the future, see #173.
